pkg/services/provisioning/dashboards/file_reader.go

```diff
@@ -412,42 +413,139 @@ type dashboardIdentity struct {
-		uidUsage:             map[string]uint8{},
-		titleUsage:           map[dashboardIdentity]uint8{},
+func newUsageTracker() *usageTracker {
+	return &usageTracker{
+		uidUsage:   map[string]uint8{},
+		titleUsage: map[dashboardIdentity]uint8{},
 	}
 }
 
-type provisioningSanityChecker struct {
-	provisioningProvider string
-	uidUsage             map[string]uint8
-	titleUsage           map[dashboardIdentity]uint8
+type usageTracker struct {
+	uidUsage   map[string]uint8
+	titleUsage map[dashboardIdentity]uint8
 }
 
-func (checker provisioningSanityChecker) track(pm provisioningMetadata) {
+func (t *usageTracker) track(pm provisioningMetadata) {
 	if len(pm.uid) > 0 {
-		checker.uidUsage[pm.uid]++
+		t.uidUsage[pm.uid]++
 	}
 	if pm.identity.Exists() {
-		checker.titleUsage[pm.identity]++
+		t.titleUsage[pm.identity]++
 	}
 }
 
-func (checker provisioningSanityChecker) logWarnings(log log.Logger) {
-	for uid, times := range checker.uidUsage {
-		if times > 1 {
-			log.Error("the same 'uid' is used more than once", "uid", uid, "provider", checker.provisioningProvider)
+type set struct {
+	data map[string]struct{}
+}
+
+func newSet() *set {
+	return &set{make(map[string]struct{})}
+}
+
+func (s *set) Add(entry string) {
+	s.data[entry] = struct{}{}
+}
+
+func (s *set) AsSlice() []string {
+	entries := make([]string, len(s.data))
+	index := 0
+
+	for entry := range s.data {
+		entries[index] = entry
+		index++
+	}
+
+	return entries
+}
+
+type duplicate struct {
+	Readers *set
+	Sum     uint8
+}
+
+func newDuplicate() *duplicate {
+	return &duplicate{Readers: newSet()}
+}
+
+type duplicateEntries struct {
+	Title map[dashboardIdentity]*duplicate
+	UID   map[string]*duplicate
+}
+
+func newDuplicateEntries() *duplicateEntries {
+	return &duplicateEntries{Title: make(map[dashboardIdentity]*duplicate), UID: make(map[string]*duplicate)}
+}
+
+type sanityChecker struct {
+	mux      sync.RWMutex
+	trackers map[string]*usageTracker
+}
+
+func newSanityChecker() *sanityChecker {
+	return &sanityChecker{trackers: make(map[string]*usageTracker)}
+}
+
+func (c *sanityChecker) Set(readerName string, tracker *usageTracker) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	c.trackers[readerName] = tracker
+}
+
+func (c *sanityChecker) getDuplicates() *duplicateEntries {
+	duplicates := newDuplicateEntries()
+
+	for readerName, tracker := range c.trackers {
+		for uid, times := range tracker.uidUsage {
+			if _, ok := duplicates.UID[uid]; !ok {
+				duplicates.UID[uid] = newDuplicate()
+			}
+			duplicates.UID[uid].Sum += times
+			duplicates.UID[uid].Readers.Add(readerName)
+		}
+
+		for id, times := range tracker.titleUsage {
+			if _, ok := duplicates.Title[id]; !ok {
+				duplicates.Title[id] = newDuplicate()
+			}
+			duplicates.Title[id].Sum += times
+			duplicates.Title[id].Readers.Add(readerName)
+		}
+	}
+
+	return duplicates
+}
+
+func (c *sanityChecker) logWarnings(log log.Logger) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	duplicates := c.getDuplicates()
```

Comment:
@sakjur Might we want to keep also the `defer c.mux.RUnlock()` call just in case we either forget to unlock after `c.getDuplicates()` or the latter panics (and gets recovered)?

Reply:
@aknuds1 That's a good point, moving the mutex logic into `.getDuplicates()` should be straightforward enough 🙂

Reply:
@sakjur An even better point :D

Reply:
I have moved lock closer to the data it protects (to the FileReader structure).